zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/callback/ZkAsyncCallbacks.java

```diff
@@ -289,7 +290,11 @@ protected boolean needRetry(int rc) {
           return false;
         }
       } catch (ClassCastException | NullPointerException ex) {
-        LOG.error("Failed to handle unknown return code {}. Skip retrying.", rc, ex);
+        if(rc == ZK_SESSION_MISMATCHED_CODE) {
+          LOG.error("Actual session ID doesn't match with expected session ID. Skip retrying.");
+        } else {
+          LOG.error("Failed to handle unknown return code {}. Skip retrying.", rc, ex);
+        }
```

Comment:
This code seems just one-off and doesn't look to extendable. Do you think this would be a good opportunity to move all Helix Project's `zookeeper-api` error codes into one class so it's easier to track?

For example, see the example in `KeeperException` (this is part of the ZooKeeper project):

```
  @Public
  public static enum Code implements KeeperException.CodeDeprecated {
    OK(0),
    SYSTEMERROR(-1),
    RUNTIMEINCONSISTENCY(-2),
    DATAINCONSISTENCY(-3),
    CONNECTIONLOSS(-4),
    MARSHALLINGERROR(-5),
    UNIMPLEMENTED(-6),
    OPERATIONTIMEOUT(-7),
    BADARGUMENTS(-8),
    APIERROR(-100),
    NONODE(-101),
    NOAUTH(-102),
    BADVERSION(-103),
    NOCHILDRENFOREPHEMERALS(-108),
    NODEEXISTS(-110),
    NOTEMPTY(-111),
    SESSIONEXPIRED(-112),
    INVALIDCALLBACK(-113),
    INVALIDACL(-114),
    AUTHFAILED(-115),
    SESSIONMOVED(-118),
    NOTREADONLY(-119);
```

Reply:
This makes sense, I'll move all of the error codes into one class.

Reply:
Based on the different exception classes that extend `ZkException` in the package `org.apache.helix.zookeeper.zkclient.exception`, only `ZkSessionMismatchedException` and `ZkInterruptedException` don't already have corresponding codes in `KeeperException.Code`, and `ZkInterruptedException` seems to be only adding thread interruption capability to `ZkException`. So it seems to me that `ZkSessionMismatchedException` is the only exception type in Helix project without a corresponding `KeeperException.Code`. Do we expect more error codes to be added in the future? Because otherwise, it seems to me that adding a new class for this one error code is a bit of an overkill. But if you think this is the right approach, I will add the class.